File: projecteuler/prob00001.hs
{-
Multiples of 3 and 5
Problem 1

If we list all the natural numbers below 10 that are multiples of 3 or 5,
we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.
-}

-- どこまで取るか
maxNumber = 999

-- 3 か 5 の倍数を取り出す
takeNumberOfMultiplesOfThreeAndFive :: Int -> Bool
takeNumberOfMultiplesOfThreeAndFive n =
  mod n 3 == 0 || mod n 5 == 0

-- 入力した整数までの 3 か 5 の倍数からなる数のリストを作る
getNumbers :: Int -> [Int]
getNumbers n = filter takeNumberOfMultiplesOfThreeAndFive [1..n]

-- 数のリストを取ってその全ての和を取る
takeSums :: [Int] -> Int
takeSums lst = foldr (+) 0 lst


main = do
  print $ takeNumberOfMultiplesOfThreeAndFive 3
  print $ takeNumberOfMultiplesOfThreeAndFive 5
  print $ takeNumberOfMultiplesOfThreeAndFive 15
  print $ (getNumbers 2 == [])
  print $ (getNumbers 3 == [3])
  print $ (getNumbers 3 == [3])
  print $ (getNumbers 5 == [3, 5])
  print $ (getNumbers 5 == [3, 5])
  print $ (getNumbers 15 == [3, 5, 6, 9, 10, 12, 15])
  print $ (takeSums [] == 0)
  print $ (takeSums [1] == 1)
  print $ (takeSums [1, 2] == 3)
  print $ (takeSums [1, 2, 3] == 6)
  print $ getNumbers maxNumber
  print $ takeSums $ getNumbers maxNumber

File: projecteuler/prob00001.py
# coding: utf-8
"""
Multiples of 3 and 5
Problem 1

If we list all the natural numbers below 10 that are multiples of 3 or 5,
we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.
"""

# どこまで取るか
MAX_NUM = 999

# 0 以外の 3 か 5 の倍数を取り出す
def is_multiple_3_or_5(n):
    if n == 0:
        return False
    else:
        return n % 3 == 0 or n % 5 == 0

# 入力した整数までの 3 か 5 の倍数からなる数のリストを作る
def get_nums(n):
    #return list(filter(is_multiple_3_or_5, list(range(n+1))))
    return [x for x in range(n+1) if is_multiple_3_or_5(x)]

# 数のリストを取ってその全ての和を取る
def takeSum(nums):
    return sum(nums)

if __name__ == "__main__":
    print(is_multiple_3_or_5(1) == False)
    print(is_multiple_3_or_5(3) == True)
    print(is_multiple_3_or_5(5) == True)
    print(is_multiple_3_or_5(15) == True)
    print(get_nums(1) == [])
    print(get_nums(3) == [3])
    print(get_nums(5) == [3,5])
    print(get_nums(15) == [3,5,6,9,10,12,15])
    print(sum(get_nums(3)) == 3)
    print(sum(get_nums(5)) == 8)
    print(sum(get_nums(5)) == 8)
    print("\nThe following number is the answer!")
    print(sum(get_nums(MAX_NUM)))
